src/api/users/mutations/loginUser/index.js

```diff
@@ -0,0 +1,71 @@
+// @flow
+import bcrypt from 'bcrypt'
+import type { ValidationError } from '@types/ValidationError'
+import createToken from '@models/Token/createToken'
+import retrieveToken from '@models/Token/retrieveToken'
+import User from '@models/User'
+import {
+  AUTHENTICATION_ERROR_INVALID_CREDENTIALS,
+  ERROR_MISSING_FIELD
+} from '@constants/errors'
+
+export type LoginUserArgs = {
+  email: string,
+  password: string
+}
+
+type LoginUserResponse = {
+  token?: string,
+  errors: Array<ValidationError>
+}
+
+const resolver = async (
+  _: any,
+  { email, password }: LoginUserArgs
+): Promise<LoginUserResponse> => {
+  const errors: Array<ValidationError> = []
+
+  if (!email || !password) {
+    errors.push({
+      field: !email ? 'email' : 'password',
+      message: ERROR_MISSING_FIELD
+    })
+
+    return { status: 'FAILURE', errors }
+  }
+
+  const user: User = await User.query().findOne({ email })
+
+  if (!user) {
+    errors.push({
+      field: 'generic',
+      message: AUTHENTICATION_ERROR_INVALID_CREDENTIALS
+    })
+
+    return { status: 'FAILURE', errors }
+  }
+
+  const isPasswordMatching: Boolean = await bcrypt.compare(
+    password,
+    user.passwordHash
+  )
+
+  if (!isPasswordMatching) {
+    errors.push({
+      field: 'generic',
+      message: AUTHENTICATION_ERROR_INVALID_CREDENTIALS
+    })
+
+    return { status: 'FAILURE', errors }
+  }
```

Comment:
I think we should do `if (!user || !isPasswordMatching) { … }`  line 39 since we send the same error anyway.

Reply:
We can't get `isPasswordMatching` if `user` is `undefined` unfortunately

Reply:
Do you have an idea to work around this ? We could wrap `isPAsswordMatching` in a try/catch, wdyt ?

Reply:
Ah. Yeah. 😄 
Maybe you could do:
```js
const isPasswordMatching: Boolean = await bcrypt.compare(
  password,
  user && user.passwordHash // Will just pass undefined if `user` is not defined.
)
```